plonk-core/src/proof_system/prover.rs

```diff
@@ -177,6 +177,12 @@ where
         // Commitments
         let mut transcript = self.preprocessed_transcript.clone();
 
+        // Append Public Inputs to the transcript
+        let pi_dense = self.cs.construct_dense_pi_vec();
+        for val in &pi_dense {
+            transcript.append(b"pi", val)
+        }
```

Comment:
I don't think we should append the dense repr, we should do the sparse (in case it doesn't affect the security) as the zeros will consume an insane amount of time.

Reply:
We could do something like this too
``` rust
for (key, val) in self.cs.public_inputs_sparse_store.iter() {
    let static_label = Box::leak(format!("pi{}", key).into_boxed_str());
    transcript.append(static_label.as_bytes(), val)
}
```

Reply:
Well, while this works, can't we use `String` or similar?? Box leaking works but it's something better to avoid if possible.

Reply:
Can we serialize the sparse representation and append to the transcript as bytes?

Reply:
1. I think we should make a PR to `merlin` so that the label doesn't need `'static` lifetime (I'm sure there was some good original reason for this, but it shouldn't be mandatory, because if the data is consumed by the hasher, it doesn't need to live that long)

2. If transcript performance is an issue, then maybe also make a PR to `merlin` to allow for Blake2b or other hashes (it looks like it might use Keccak right now...probably not the best choice for performance). The transcript also has to support Poseidon or other hashes if recursion will ever be used (as part of the transcript is verified in-circuit)

3. I think it should be fine to append a sparse repr instead of dense, but it has to be done in the correct way. The most important thing is that there is only one way to represent each PI in the transcript, and no two PI share the same repr in the transcript. As @davidnevadoc points out this requires that the sparse repr does not have any zero values.

Reply:
I'm AFK or else I would just check this now...instead I'll leave a comment so I don't forget to mention it: has anyone checked PlookUp branch to see if the lookup table vectors get included in the transcript?

Reply:
I agree with these suggestions but since `merlin` has not merged a PR in more than a year maybe we should create our own fork.

Reply:
Something to be wary of is iteration order. If you aren't using an ordered hash map then there are no guarantees on how the pi are returned from the map. You should use either an ordered map or perform a sort by index (public_input_indices) before inserting into the transcript.

Reply:
Finally went for a `BTreeMap` struct inside `PI` to ensure the values are returned  in order when appended to the transcript.

Reply:
> I think we should make a PR to merlin so that the label doesn't need 'static lifetime (I'm sure there was some good original reason for this, but it shouldn't be mandatory, because if the data is consumed by the hasher, it doesn't need to live that long)

Ah, no, actually, the `'static` restriction is there on purpose to prevent exactly this type of (mis)use! 

The labels are supposed to be fixed, rather than input-dependent, because the labels are used for message framing for the data that's fed into the transcript.  Merlin uses a tag-length-value format, so the transcript is guaranteed to be parseable (and hence produce a non-malleable transcript hash) if the labels are all fixed and none of them are prefixes of each other.

In this case, the correct way to feed variable data into the transcript is to use two messages, one for the application-level label, and one for the application-level data, each with their own `'static` label at the transcript level.

> I agree with these suggestions but since merlin has not merged a PR in more than a year maybe we should create our own fork.

The upstream moved to https://github.com/zkcrypto/merlin/ ; I'm happy to merge PRs there, but not this one, since the design is trying to steer you away from doing this :)

Reply:
Ah, this is understandable (I hadn't realized the use of `'static` was connected to this property). You have a good point that the variable-length PIs have to be handled better here - this kind of thing is a bit of a challenge (no pun) because of circuit configurability at runtime.

However I still have doubts whether using `'static` for this purpose is really the best way to accomplish this. For example, the `Box::leak` code above causes this guarantee to break, without any indication at all that something went wrong (other than leaking memory - but the memory guarantee is now tied to the correctness). To get rustc to enforce these guarantees at compile time, perhaps something like enforcing the labels to be from an `enum` might be more effective.

Reply:
Because the actual circuit structure (and the entire proof system actually) can be configured at runtime, the notion of "compile-time" labels is just not correct. In the case of this repo, the keys are fixed at the _circuit compile time_ but not at the _Rust runtime_. The only way a safety feature would be useful here is if you could constrain the labels to be fixed during circuit compilation.

Reply:
> For example, the Box::leak code above causes this guarantee to break, without any indication at all that something went wrong (other than leaking memory - but the memory guarantee is now tied to the correctness). 

Uh, sure, I guess, but the use of `Box::leak` *in a loop over user-supplied data* to intentionally leak memory in the middle of a crypto implementation seems like a pretty clear indicator that something is wrong with the implementation, no?

> Because the actual circuit structure (and the entire proof system actually) can be configured at runtime, the notion of "compile-time" labels is just not correct; it is defined at the wrong abstraction level.

No, it isn't. You can of course write an application that doesn't include all of its data in its transcript, or does so incorrectly (which is how we ended up here). There's no way for the transcript construction to solve that for you. What it can, and does, solve for you, is ensure that you only have to worry about the application-level data, and not the low-level details of ensuring that that data is unambiguously encoded in the transcript hash.

Reply:
> No, it isn't. You can of course write an application that doesn't include all of its data in its transcript, or does so incorrectly (which is how we ended up here). There's no way for the transcript construction to solve that for you. What it can, and does, solve for you, is ensure that you only have to worry about the application-level data, and not the low-level details of ensuring that that data is unambiguously encoded in the transcript hash.

It's the fact that the labels are `'static` that's the problem. Which labels to use when specifying the transcript protocol do not need to be known at _Rust compile time_ to be correct. Handling the "low-level details" in this case needs to be done by this repo; the unambiguous encoding _is defined at Rust runtime_.

Reply:
> Uh, sure, I guess, but the use of `Box::leak` _in a loop over user-supplied data_ to intentionally leak memory in the middle of a crypto implementation seems like a pretty clear indicator that something is wrong with the implementation, no?

I agree it's an implicit indication - and in this case it did work exactly the way you intended, because none of us would have accepted code containing `Box::leak` as a good idea - but it also isn't explicit, in the sense that using `Box::leak` is entirely memory-safe and type-safe, so it is entirely possible to violate the API requirements at compile-time without violating any of `rustc`'s compile-time rules.

At minimum, I didn't initially understand the purpose of using `'static` here, and didn't recognize the functional consequences of `Box::leak`. It's possible I'm just an idiot, but in the unlikely event I'm not, it would be nice if it's possible to enforce the API requirements using the type system instead of the borrow checker, which would offer a stronger guarantee, because it could be more explicit, and also not-workaround-able. However, I'll admit that it's not obvious to me how to do it.

> It's the fact that the labels are 'static that's the problem. The labels used when specifying the transcript protocol do not need to be known at Rust compile time to be correct. Handling the "low-level details" in this case needs to be done by this repo; the unambiguous encoding is defined at Rust runtime.

I think the `'static` discussion was a bit of a red herring (sorry!) and it's better to leave `'static` as it is (unless it's replaced with something better). The issue is we have multiple levels of abstraction here, with each layer having it's own requirements and guarantees.

Specifically, we have `merlin` offering certain guarantees about the transcript, as long as certain requirements are met (as described by @hdevalence above and the `merlin` docs). Plonk also needs to offer similar guarantees to the users of the proving system (circuit designers, etc) as long as its own requirements are met. In the former case, we can ensure the requirements are met at `rustc` compile-time, whereas the latter needs to ensure they are met at run-time. 

As Henry said above, there could be a fixed label set for Plonk at compile time (thus meeting the `'static` requirement) and any circuit-level configuration is handled as application-level data. For example, there is one fixed label "pi" for public inputs and then the data for "pi" is itself a run-time created transcript of some kind encoding the structure of the inputs. 

Now, there is a theoretical issue, that if the application-level itself is trying to offer the same guarantees and generality as `merlin` does, then duplicating this at two layers of abstraction is at best wasteful and at worst doubling the vulnerability space. In this general case I think @bhgomes is probably correct that everything must be enforced at run-time anyway. However, at least in this initial case (for example, variable length public inputs) this is probably not necessary.